File: 146.lru-缓存.java
import java.util.LinkedHashMap;

/*
 * @lc app=leetcode.cn id=146 lang=java
 *
 * [146] LRU 缓存
 */

// @lc code=start
class LRUCache {
    int capacity;
    LinkedHashMap<Integer, Integer> cache = new LinkedHashMap<>();

    public LRUCache(int capacity) {
        this.capacity = capacity;
    }

    public int get(int key) {
        // 不存在返回-1
        if (!cache.containsKey(key)) {
            return -1;
        }
        // 提升为最近使用
        makeRecently(key);
        return cache.get(key);
    }

    public void put(int key, int value) {
        // 包含键，提升为最近使用，赋值
        if (cache.containsKey(key)) {
            makeRecently(key);
            cache.put(key, value);
            return;
        }
        // 大于空间，淘汰最久使用键，然后赋新的键值
        if (cache.size() >= capacity) {
            int oldKey = cache.keySet().iterator().next();
            cache.remove(oldKey);
        }
        cache.put(key, value);
    }

    // 将数据提升为最近使用
    public void makeRecently(int key) {
        if (cache.containsKey(key)) {
            int val = cache.get(key);
            cache.remove(key);
            cache.put(key, val);
        }
    }

}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */
// @lc code=end

File: python/146.lru-缓存.py
#
# @lc app=leetcode.cn id=146 lang=python
#
# [146] LRU 缓存
#

# @lc code=start
from collections import OrderedDict
class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.cache = OrderedDict()
        

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key in self.cache:
            self.cache.move_to_end(key)
            return self.cache[key]
        else:
            return -1
        

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        self.cache.add(key, value)
        self.cache.move_to_end(key)
        
        if self.cache.keys > self.capacity:
            self.cache.popitem(last=False)
        


# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)
# @lc code=end
obj = LRUCache(2)
obj.put(1, 1)
obj.put(2, 2)
print(obj.get(1))    # 返回 1
obj.put(3, 3)        # 该操作会使得密钥 2 作废
print(obj.get(2))    # 返回 -1 (未找到)
obj.put(4, 4)        # 该操作会使得密钥 1 作废
print(obj.get(1))    # 返回 -1 (未找到)
print(obj.get(3))    # 返回 3
print(obj.get(4))    # 返回 4
